File: prefix_sum.java
public class PrefixSum {
    public static int[] calculatePrefix(int[] arr) {
        int[] prefix = new int[arr.length];
        prefix[0] = arr[0];
        for(int i=1; i<arr.length; i++) {
            prefix[i] = prefix[i-1] + arr[i];
        }
        return prefix;
    }

    public static int rangeSum(int[] prefix, int l, int r) {
        return prefix[r] - (l > 0 ? prefix[l-1] : 0);
    }
    
    // Driver code
    public static void main(String[] args) {
        int[] arr = {1, 3, 5, 7};
        int[] ps = calculatePrefix(arr);
        System.out.println(Arrays.toString(ps));  // [1, 4, 9, 16]
    }
}

File: prefix_sum.py
def prefix_sum(arr):
    n = len(arr)
    prefix = [0] * n
    prefix[0] = arr[0]
    for i in range(1, n):
        prefix[i] = prefix[i-1] + arr[i]
    return prefix

def range_sum(prefix, l, r):
    return prefix[r] - (prefix[l-1] if l > 0 else 0)

# Example Usage
arr = [2, 4, 6, 8, 10]
ps = prefix_sum(arr)
print(ps)           # [2, 6, 12, 20, 30]
print(range_sum(ps, 1, 3))  # 18
